File: python/src/services/round_evaluator.py
from __future__ import annotations

from dataclasses import dataclass, field

from src.models.contracts import MappedPlayerState
from src.models.enums import CellState, PlayerTrackingState
from src.models.game_session_model import GameSessionModel


@dataclass(slots=True)
class EvaluationResult:
    survivors: list[str] = field(default_factory=list)
    eliminated: list[str] = field(default_factory=list)
    reason_map: dict[str, str] = field(default_factory=dict)


class RoundEvaluator:
    def evaluate(
        self,
        session: GameSessionModel,
        mapped_players: list[MappedPlayerState],
    ) -> EvaluationResult:
        players_by_id = {player.player_id: player for player in mapped_players}
        result = EvaluationResult()

        for participant_id in session.round_state.participant_ids:
            mapped_player = players_by_id.get(participant_id)
            session_player = session.players[participant_id]

            eliminated, reason = self._check_elimination(
                mapped_player, session_player.tracking_state, session.grid.cell_states
            )

            if eliminated:
                result.eliminated.append(participant_id)
                result.reason_map[participant_id] = reason
            else:
                result.survivors.append(participant_id)

        return result

    @staticmethod
    def _check_elimination(
        mapped_player: MappedPlayerState | None,
        tracking_state: PlayerTrackingState,
        cell_states: dict[tuple[int, int], CellState],
    ) -> tuple[bool, str]:
        if mapped_player is None:
            return True, "not found in mapped players"
        if tracking_state is PlayerTrackingState.MISSING:
            return True, "player missing"
        if not mapped_player.in_bounds:
            return True, "out of bounds"
        if mapped_player.occupied_cell is None:
            return True, "no occupied cell"
        if mapped_player.ambiguous:
            return True, "ambiguous position"
        if cell_states.get(mapped_player.occupied_cell) is CellState.RED:
            return True, "standing on red cell"
        return False, ""

File: cpp/src/services/round_evaluator.cpp
#include "services/round_evaluator.h"

namespace sag {

EvaluationResult RoundEvaluator::evaluate(
    const GameSessionModel& session,
    const std::vector<MappedPlayerState>& mapped_players) const
{
    std::unordered_map<std::string, const MappedPlayerState*> players_by_id;
    for (auto& mp : mapped_players)
        players_by_id[mp.player_id] = &mp;

    EvaluationResult result;

    for (auto& participant_id : session.round_state.participant_ids) {
        const MappedPlayerState* mapped = nullptr;
        if (auto it = players_by_id.find(participant_id); it != players_by_id.end())
            mapped = it->second;

        auto session_it = session.players.find(participant_id);
        PlayerTrackingState tracking = (session_it != session.players.end())
            ? session_it->second.tracking_state
            : PlayerTrackingState::Missing;

        auto check = check_elimination(mapped, tracking, session.grid.cell_states);

        if (check.eliminated) {
            result.eliminated.push_back(participant_id);
            result.reason_map[participant_id] = std::move(check.reason);
        } else {
            result.survivors.push_back(participant_id);
        }
    }

    return result;
}

RoundEvaluator::EliminationCheck RoundEvaluator::check_elimination(
    const MappedPlayerState* mapped_player,
    PlayerTrackingState tracking_state,
    const CellStateMap& cell_states)
{
    if (mapped_player == nullptr)
        return {true, "not found in mapped players"};

    if (tracking_state == PlayerTrackingState::Missing)
        return {true, "player missing"};

    if (!mapped_player->in_bounds)
        return {true, "out of bounds"};

    if (!mapped_player->occupied_cell.has_value())
        return {true, "no occupied cell"};

    if (mapped_player->ambiguous)
        return {true, "ambiguous position"};

    auto it = cell_states.find(*mapped_player->occupied_cell);
    if (it != cell_states.end() && it->second == CellState::Red)
        return {true, "standing on red cell"};

    return {false, ""};
}

} // namespace sag
